File: control.py
import numpy as np



def wrapToPi(theta):
    while theta < -np.pi:
        theta = theta + 2*np.pi
    while theta > np.pi:
        theta = theta - 2*np.pi
    return theta

def visualServoingCtl (camera, desiredState, actualState, v_des):         #camera里面的参数应该都是m（由handler的157行）
    # specifiy the acutal state for better readability
    x = actualState[0]           # 按理来说是a
    y = actualState[1]           # 按理来说是H/2
    theta = actualState[2]       # 角度
    
    # some crazy parameters   
    lambda_x_1 = 10               #  自己设定的正系数
    lambda_w_1= 3000              #  自己设定的正系数
    lambdavec = np.array([lambda_x_1,lambda_w_1])
    
    # state if it is a row or a column controller
    controller_type = 0            # 用来选择 Lx还是Ly  ，   0就是选择Lx   表示横向控制
    
    # s_c = J * u = L_sc * T_r->c * u_r
    
    # Computation of the Interaction Matrix as presented by Cherubini & Chaumette
    # relates the control variables [v,w] in the camera frame to the change of the features [x,y,theta]
    angle = camera.tilt_angle         #相机与水平面的夹角
    delta_z = camera.deltaz          # 相机距离地面的高度
    IntMat = np.array([[(-np.sin(angle)-y*np.cos(angle))/delta_z, 0, x*(np.sin(angle)+y*np.cos(angle))/delta_z, x*y, -1-x**2,  y],
                       [0, -(np.sin(angle)+y*np.cos(angle))/delta_z, y*(np.sin(angle)+y*np.cos(angle))/delta_z, 1+y**2, -x*y, -x],
                       [np.cos(angle)*np.power(np.cos(theta),2)/delta_z, np.cos(angle)*np.cos(theta)*np.sin(theta)/delta_z, 
                        -(np.cos(angle)*np.cos(theta)*(y*np.sin(theta) + x*np.cos(theta)))/delta_z, 
                        -(y*np.sin(theta) + x*np.cos(theta))*np.cos(theta), -(y*np.sin(theta) + x*np.cos(theta))*np.sin(theta), -1]])

    # Computation of the transformation from the robot to the camera frame
    delta_y = camera.deltay                  # 相机与轮子的水平距离
    TransfMat = np.array([[0,-delta_y],
                         [-np.sin(angle),0],
                         [np.cos(angle),0],
                         [0,0],
                         [0,-np.cos(angle)],
                         [0,-np.sin(angle)]])
    Trans_vel = TransfMat[:,0]                  #  Tv
    Trans_ang = TransfMat[:,1]                  #  Tw

    # Computation of the Jacobi-Matrix for velocity and angular velocity and their pseudo inverse
    # The Jacobi-Matrix relates the control variables in the robot frame to the change of the features
    Jac = np.array([IntMat[controller_type,:],IntMat[2,:]])        # Ar 矩阵
    Jac_vel = np.matmul(Jac,Trans_vel)                  #  Ar    （matmul 矩阵相乘）
    # Jac_vel_pi = np.linalg.pinv([Jac_vel])
    Jac_ang = np.matmul(Jac,Trans_ang)                  #  Br
    Jac_ang_pi = np.linalg.pinv([Jac_ang])              #  求矩阵的伪逆
    
    # Compute the delta, in this case the difference between the actual state and the desired state
    trans_delta = actualState[controller_type] - desiredState[controller_type]   # e_x
    ang_delta = actualState[2] - desiredState[2]             # e_theta
    delta = np.array([trans_delta,wrapToPi(ang_delta)])
    
    # Compute the feedback control for the angular velocity
    temp = lambdavec * delta
    ang_fb = np.matmul(-Jac_ang_pi.T,(temp + Jac_vel * v_des))
    return ang_fb


class Camera:            # 相机的各个参数
    def __init__(self, id,deltax,deltay,deltaz,tilt_angle,height,width,focal_length,sensor_width):
        self.id = id
        self.deltax = deltax                   # Translation to the origin of the robot
        self.deltay = deltay
        self.deltaz = deltaz                    # Heigth of the camera
        self.tilt_angle = tilt_angle            # Tilt Angle
        self.height = height                    # Camera Intrinsics
        self.width = width
        self.focal_length = focal_length
        self.sensor_width = sensor_width


if __name__ == '__main__':

    camera = Camera(1,1.2,0,1,np.deg2rad(-80),0.96,0,0,1)    #相机参数
    imgWidth = 512                                           #图像的宽高
    P = [10]                                                 #下交点的坐标，像素坐标系
    Angle = np.deg2rad(-30)                                 #导航线夹角
    desiredState = np.array([0.0, imgWidth/2, 0.0])
    actualState = np.array([P[0], imgWidth/2, Angle])
    v_des = 0.2                                              #设定好的线速度  m/s


    w = visualServoingCtl(camera, desiredState, actualState, v_des)
    print(w)

File: src/control.cpp
#include "control.hpp"

float visualServoingCtl(Camera& camera, vector<float> &desiredState, vector<float> &actualState, float v_des){
    // specifiy the acutal state for better readability
    // cout << "x: " << actualState[0] << endl;
    // float x = actualState[0]*1.34/1.37/1000;
    float x = actualState[0];
    float y = actualState[1];
    float theta = actualState[2];

    // some crazy parameters
    float lambda_x_1 = 1;                     // 正系数1
    float lambda_w_1= 0.1;                     // 正系数2
    Eigen::Vector2f lambdavec(lambda_x_1, lambda_w_1);

    // state if it is a row or a column controller
    int controller_type = 0;                    //横向控制

    float angle = camera.tilt_angle;            // 相机与水平面的夹角
    float delta_z = camera.deltaz;              // 相机距离地面的高度
    
    Eigen::Matrix<float, 3, 6> IntMat;
    // IntMat << (-sin(angle)-y*cos(angle))/delta_z, 0, x*(sin(angle)+y*cos(angle))/delta_z, x*y, -1-pow(x, 2),  y,
    //           0, -(sin(angle)+y*cos(angle))/delta_z, y*(sin(angle)+y*cos(angle))/delta_z, 1+pow(y, 2), -x*y, -x,
    //           cos(angle)*pow(cos(theta),2)/delta_z, cos(angle)*cos(theta)*sin(theta)/delta_z, 
    //           -(cos(angle)*cos(theta)*(y*sin(theta) + x*cos(theta)))/delta_z, 
    //           -(y*sin(theta) + x*cos(theta))*cos(theta), -(y*sin(theta) + x*cos(theta))*sin(theta), -1;

    double f = 4.11e-3;
    IntMat << f*(-sin(angle)-y/f*cos(angle))/delta_z, 0, x*(sin(angle)+y/f*cos(angle))/delta_z, x*y/f, (-f*f-pow(x, 2))/f,  y,
              0, -(sin(angle)+y*cos(angle))/delta_z, y*(sin(angle)+y*cos(angle))/delta_z, 1+pow(y, 2), -x*y, -x,
              cos(angle)*pow(cos(theta),2)/delta_z, cos(angle)*cos(theta)*sin(theta)/delta_z, 
              -(cos(angle)*cos(theta)*(y*sin(theta) + x*cos(theta)))/delta_z/f, 
              -(y*sin(theta) + x*cos(theta))*cos(theta)/f, -(y*sin(theta) + x*cos(theta))*sin(theta)/f, -1;
    
    float delta_y = camera.deltay;
    Eigen::Matrix<float, 6, 2> TransfMat;
    TransfMat << 0,-delta_y,
                -sin(angle),0,
                cos(angle),0,
                0,0,
                0,-cos(angle),
                0,-sin(angle);

    Eigen::Vector<float, 6> Trans_vel = TransfMat.col(0);                  //  Tv
    Eigen::Vector<float, 6> Trans_ang = TransfMat.col(1);                  //  Tw

    Eigen::Matrix<float, 2, 6> Jac; 
    Jac << IntMat.row(0), IntMat.row(2);
    Eigen::Vector2f Jac_vel = Jac*Trans_vel;

    Eigen::Vector2f Jac_ang = Jac*Trans_ang;
    auto m = pseudoInverse(Jac_ang);
    Eigen::Vector2f Jac_ang_pi(m(0), m(1));
    // cout << "Jac_ang_pi.shape: " << Jac_ang_pi.size() << endl;

    float trans_delta = actualState[controller_type] - desiredState[controller_type];    // e_x
    float ang_delta = actualState[2] - desiredState[2];                                  // e_theta
    Eigen::Vector2f delta(trans_delta, wrapToPi(ang_delta));

    Eigen::Vector2f temp = lambdavec.cwiseProduct(delta);
    float ang_fb = -Jac_ang_pi.transpose()*(temp + Jac_vel * v_des);

    return ang_fb;
}

float getWheelAngle(float w_robot, float v_des, float L, float B){
  double tmp = v_des/abs(w_robot);
  float R = tmp > B ? tmp : B + 0.0001; //R必须大于B
  // float R = tmp;
  float theta_1 = atan(L/(R-B)); //内轮
  float theta_2 = atan(L/(R+B)); //外轮
  return w_robot >= 0 ? (rad2deg(theta_1) + rad2deg(theta_2))/2 : -(rad2deg(theta_1) + rad2deg(theta_2))/2;
}

float control_unit(Camera& cam, float L, float B, float frame_height, float v_des, float e_x, float e_angle){

  // cam = Camera(1.2, 0, 1, deg2rad(-80));
  // frame_height = 512;                // 像素画面高度

  // v_des = 0.2;                      // 设定好的线速度  m/s
  // e_x= 10;                          // 横向偏差
  // e_angle = -30;            // 角偏差

  float dx = 3e-6;
  e_x = e_x*dx; // e_x转换到图像坐标系中, 100像素距离 == 0.3mm图像坐标系上的距离

  vector<float> desiredState{0.0, frame_height/2*dx, 0.0};
  vector<float> actualState{e_x, frame_height/2*dx, deg2rad(e_angle)};

  float w = visualServoingCtl(cam, desiredState, actualState, v_des);
  cout << "机器人的反馈角速度: " << w << "rad/s" << endl;

  float wheelAngle = getWheelAngle(w, v_des, L, B);
  // cout << "车轮转角：" << wheelAngle << endl;

  return wheelAngle;
}

string angle2signal(SerialPort *serialPort, float wheelAngle){
  // float prop = 39 / 90.f;
  // int m = int(-wheelAngle*prop) + 39;
  // int n = m*(144.0 - 112.0)/78 + 112;
  // string s = DecIntToHexStr(n);
  // if(s.length() < 2){
  //   s = "0" + s;
  // }
  // return "FF010125" + s + "00000000FF";
  // wheelAngle = -wheelAngle;
  // if(wheelAngle < -30){
  //   return "FF010155D00000FEFF";
  // }
  // else if(wheelAngle < 0){
  //   return "FF010155B00000FEFF";
  // }
  // else if(wheelAngle < 30){
  //   return "FF010155500000FEFF";
  // }
  // else{
  //   return "FF010155300000FEFF";
  // }

  // 70 78 | 88 90  原本速度67

  wheelAngle = -wheelAngle;

  if(wheelAngle < 0){
    if(wheelAngle > -60)
      return "FF010172880000FEFF";
    else
      return "FF010172900000FEFF";
  }
  else{
    if(wheelAngle < 70)
      return "FF010172780000FEFF";
    else
      return "FF010172700000FEFF";
  }
}